core/src/test/java/hudson/util/AtomicFileWriterTest.java

```diff
@@ -124,4 +142,72 @@ public void badPath() throws Exception {
                        containsString("exists and is neither a directory nor a symlink to a directory"));
         }
     }
+
+    @Issue("JENKINS-48407")
+    @Test
+    public void preserveSpecificPermissions() throws IOException, InterruptedException {
```

Comment:
this is actually changing existing behaviour is it not.  Before the permissions would have been whatever we created the new file as (as it was create new file, delete old file, rename new file).

I think this is possibly bad behaviour and we should not try and do this.

Reply:
I initially thought preserving permissions would be a good idea but now I agree with @jtnord. I believe this would also cause files who got new permissions after #2548 to keep the same bad permissions for anyone who is running the latest weekly builds.

Reply:
I agree I'm not keeping the following behaviour, but it was on purpose as explained in [my comment](https://issues.jenkins-ci.org/browse/JENKINS-48407?focusedCommentId=322043&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-322043).

So, though I obviously agree this is not preserving the previous behaviour, I still think this is the right thing to do:
* preserve the previous perms, if the file was already existing
* use the default permissions, iow respecting `umask` (if we can, sigh, does not seem so easy from Java code...), and respecting **for that case** the previous behaviour.

Now, from reading below and so on, does not look like it'd be so easy, especially as on Windows that permission retrieval seems to barf... WAT.